src/main/java/emu/grasscutter/game/entity/EntityAvatar.java

```diff
@@ -175,12 +175,28 @@ public void addEnergy(float amount, PropChangeReason reason, boolean isFlat) {
 		
 		// Set energy and notify.
 		if (newEnergy != curEnergy) {
-			this.avatar.setCurrentEnergy(curEnergyProp, newEnergy);
-			
-			this.getScene().broadcastPacket(new PacketAvatarFightPropUpdateNotify(this.getAvatar(), curEnergyProp));
-			this.getScene().broadcastPacket(new PacketEntityFightPropChangeReasonNotify(this, curEnergyProp, newEnergy, reason));
+			this.setCurrentEnergy(curEnergyProp,newEnergy,reason);
+		}
+	}
+
+	private void setCurrentEnergy() {
+		float curEnergy;
+		FightProperty curEnergyProp = this.getAvatar().getSkillDepot().getElementType().getCurEnergyProp();
+		FightProperty maxEnergyProp = this.getAvatar().getSkillDepot().getElementType().getMaxEnergyProp();
+		PropChangeReason reason=PropChangeReason.PROP_CHANGE_REASON_ABILITY;
+		if(this.getPlayer().getEnergyUsage()){
+			curEnergy=this.avatar.getCurrentEnergy();
+		}else{
+			curEnergy= this.getFightProperty(maxEnergyProp);
 		}
-	} 
+		setCurrentEnergy(curEnergyProp,curEnergy,reason);
+	}
+
+	private void setCurrentEnergy(FightProperty curEnergyProp,float newEnergy,PropChangeReason reason) {
+		this.avatar.setCurrentEnergy(curEnergyProp, newEnergy);
+		this.getScene().broadcastPacket(new PacketAvatarFightPropUpdateNotify(this.getAvatar(), curEnergyProp));
+		this.getScene().broadcastPacket(new PacketEntityFightPropChangeReasonNotify(this, curEnergyProp, newEnergy, reason));
+	}
```

Comment:
It is confusing that there is now a `setCurrentEnergy` for both `Avatar` and `EntityAvatar`. Please unify them and put them into either one of those classes, probably in `EntityAvatar`.